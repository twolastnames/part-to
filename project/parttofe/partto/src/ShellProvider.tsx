import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
