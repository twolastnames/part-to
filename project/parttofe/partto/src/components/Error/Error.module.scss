@import "../../App.scss";

.error {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: space-between;
  flex-direction: column;
}

.code {
  font: var(--font-strong-xxxlarge);
  text-align-last: center;
  padding-top: var(--gap-xlarge);
}

.description {
  font: var(--font-regular-large);
  padding: var(--gap-xlarge);
}
