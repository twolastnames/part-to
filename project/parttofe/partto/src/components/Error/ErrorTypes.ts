export interface ErrorProps {
  code: number;
}
