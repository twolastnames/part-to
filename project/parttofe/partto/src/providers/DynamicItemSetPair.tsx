import React, {
  createContext,
  useState,
  Context,
  MutableRefObject,
  createRef,
  useContext,
  SetStateAction,
} from "react";
import { PropsWithChildren } from "react";
import { Duration, getDuration } from "../shared/duration";
import { debounce, DebouncedFunc } from "lodash";

export type DynamicItemSetPairType = {
  selected: number;
  setSelected: (selected: number) => void;
  getTotal: () => number;
  goBack: () => void;
  goForward: () => void;
  showDuration: Duration;
  togglePause: () => void;
  paused: boolean;
  nextShowDuration: () => void;
};

export const UndefinedDynamicItemSetPair = {
  selected: 0,
  setSelected: (_: number) => undefined,
  goBack: () => undefined,
  goForward: () => undefined,
  getTotal: () => 0,
  showDuration: getDuration(0),
  togglePause: () => undefined,
  paused: true,
  nextShowDuration: () => undefined,
};

export function useSingleOfPair({ context }: ContextDescription) {
  return useContext(context) || UndefinedDynamicItemSetPair;
}

type ContextType = Context<undefined | DynamicItemSetPairType>;

export type ContextDescription = {
  context: ContextType;
  setCount: (value: number) => void;
  setSettingKey: (settingKey: string) => void;
  setDurations: (value: Array<Duration>) => void;
};

const defaultDurations = [1000, 2000, 4000, 8000];

const getDurations = (key: string) =>
  JSON.parse(
    localStorage.getItem(`${key}Durations`) || JSON.stringify(defaultDurations),
  ).map(getDuration);

const setDurations = (key: string, value: Array<Duration>) => {
  localStorage.setItem(
    `${key}Durations`,
    JSON.stringify(value.map((duration) => duration.toMilliseconds())),
  );
};

const leftCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const leftSettingKey: MutableRefObject<string | null> = createRef();
const getLeftSettingKey = () => leftSettingKey.current || "left";
function setLeftCount(value: number) {
  leftCount.current && leftCount.current(value);
}
const LeftDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const LeftContext: ContextDescription = {
  context: LeftDynamicItemSetPairContext,
  setCount: setLeftCount,
  setSettingKey: (value: string) => {
    leftSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getLeftSettingKey(), value);
  },
};

const rightCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const rightSettingKey: MutableRefObject<string | null> = createRef();
const getRightSettingKey = () => leftSettingKey.current || "left";

function setRightCount(value: number) {
  rightCount.current && rightCount.current(value);
}
const RightDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const RightContext: ContextDescription = {
  context: RightDynamicItemSetPairContext,
  setCount: setRightCount,
  setSettingKey: (value: string) => {
    rightSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getRightSettingKey(), value);
  },
};

type GoADirection = (
  totalPages: number,
  setSelected: (mutator: (arg: ProviderState) => ProviderState) => void,
) => void;

const goBack: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected <= 0 ? length - 1 : last.selected - 1,
  }));
};

const goForward: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected >= length - 1 ? 0 : last.selected + 1,
  }));
};

type ProviderState = {
  selected: number;
  showDuration: number;
  paused: boolean;
  count: number;
};

function useProvider({
  setCount,
  getSettingKey,
  togglePauseClickSlower,
}: {
  getSettingKey: () => string;
  setCount: MutableRefObject<((value: number) => void) | null>;
  togglePauseClickSlower: DebouncedFunc<
    (setState: (arg: SetStateAction<ProviderState>) => void) => void
  >;
}) {
  const durations = getDurations(getSettingKey());
  const [{ count, selected, showDuration, paused }, setState] =
    useState<ProviderState>({
      selected: 0,
      showDuration: Math.floor(durations.length / 2),
      paused: false,
      count: 0,
    });

  setCount.current = (value: number) => {
    if (value === count) {
      return;
    }
    setState((previous) => ({ ...previous, count: value }));
  };

  if (selected !== 0 && (selected < 0 || selected >= count)) {
    setState((current) => ({ ...current, selected: 0 }));
  }

  const duration = durations[showDuration];
  return {
    selected: selected,
    setSelected: (selected: number) => {
      setState((current) => ({ ...current, selected }));
    },
    goBack: () => {
      goBack(count, setState);
    },
    goForward: () => {
      goForward(count, setState);
    },
    showDuration: duration,
    paused: paused,
    getTotal: () => count,
    togglePause: () => {
      togglePauseClickSlower(setState);
    },
    nextShowDuration: () => {
      const next = showDuration + 1;
      setState((previous) => ({
        ...previous,
        showDuration: next >= durations.length ? 0 : next,
      }));
    },
  };
}

const leftTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function LeftDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <LeftDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: leftCount,
        getSettingKey: getLeftSettingKey,
        togglePauseClickSlower: leftTogglePauseClickSlower,
      })}
    >
      {children}
    </LeftDynamicItemSetPairContext.Provider>
  );
}

const rightTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function RightDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <RightDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: rightCount,
        getSettingKey: getRightSettingKey,
        togglePauseClickSlower: rightTogglePauseClickSlower,
      })}
    >
      {children}
    </RightDynamicItemSetPairContext.Provider>
  );
}
