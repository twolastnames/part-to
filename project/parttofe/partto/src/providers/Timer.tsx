import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { PropsWithChildren } from "react";
import { DateTime, getDateTime } from "../shared/dateTime";
import { Duration, getDuration } from "../shared/duration";
import { RunState, TaskDefinitionId } from "../api/sharedschemas";
import { Timer } from "../components/Timer/Timer";
import {
  addAlarmNote,
  removeNote,
} from "../components/Layout/NavigationBar/Noted/Noted";
import {
  addRunStateListener,
  removeRunStateListener,
} from "../shared/runStateMessage";

type TimerDescription = {
  component: ReactNode;
  duration: Duration;
  started: DateTime;
  message: string;
  enforced: boolean;
  task: TaskDefinitionId;
  setOnLocate: (callback: () => void) => void;
};

export type TimerType = {
  timers: { [id: string]: TimerDescription };
};

export const TimerContext = createContext<undefined | TimerType>(undefined);

type TaskKeyed = { task: TaskDefinitionId };

type TimerFactory<T> = (args: T) => {
  component: ReactNode;
  duration: Duration;
  task: TaskDefinitionId;
};

function getMappedTimers<T extends TaskKeyed>(
  getTimer: TimerFactory<T>,
  timers: Array<T> | undefined,
): { [id: string]: TimerDescription } {
  return (
    timers?.reduce(
      (current, timer) => ({
        ...current,
        [timer.task]: getTimer(timer),
      }),
      {},
    ) || ({} as { [id: TaskDefinitionId]: T })
  );
}

export function TimerProvider({ children }: PropsWithChildren) {
  const [runState, setRunState] = useState<RunState | undefined>();
  const offsets = useRef<{
    [id: string]: Duration;
  }>({});
  const onLocates = useRef<{ [task: string]: () => void }>({});
  const audio = useRef(new Audio(require("./messageAlert.mp3")));
  const timestamp = runState?.timestamp;
  const timers = useMemo(
    () => ({
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: true,
          message: "Duty Complete?",
          component: (
            <Timer
              start={started}
              duration={duration}
              adjustment={{
                offset: offsets.current[task] || getDuration(0),
                setOffset: (value: Duration) => (offsets.current[task] = value),
              }}
            />
          ),
        }),
        runState?.timers.enforced || [],
      ),
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: false,
          message: "Past Expected Time",
          component: <Timer start={started} duration={duration} />,
        }),
        runState?.timers.laxed || [],
      ),
      ...getMappedTimers(
        ({ task, till }) => ({
          task,
          started: timestamp,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration: till,
          enforced: true,
          message: "Need to Start Duty",
          component: <Timer start={timestamp} duration={till} />,
        }),
        runState?.timers.imminent || [],
      ),
    }),
    [runState, offsets, timestamp],
  );

  useEffect(() => {
    addRunStateListener(setRunState);
    return () => {
      removeRunStateListener(setRunState);
    };
  }, [setRunState]);

  useEffect(() => {
    const id = setInterval(() => {
      const now = getDateTime();
      for (const [key, timer] of Object.entries(timers)) {
        const expireAt = timer.started
          .add(offsets.current[key] || getDuration(0))
          .add(timer.duration);
        if (!timer.enforced || now.sinceEpoch() < expireAt.sinceEpoch()) {
          continue;
        }
        addAlarmNote({
          key,
          heading: "Alarm Expired",
          detail: timer.message,
          onClick: () => {
            onLocates.current[key]?.();
          },
        });
        if (!audio.current.paused) {
          return;
        }
        try {
          audio.current.play();
        } catch (NotAllowedError) {
          console.error(
            "could not start an alarm... likely because the page has not been interacted with",
          );
        }
        audio.current.loop = true;
        return;
      }
      for (const key of Object.keys(timers)) {
        removeNote(key);
      }
      audio.current.pause();
    }, 1000);
    return () => {
      clearInterval(id);
    };
  }, [offsets, runState, timers]);

  return (
    <TimerContext.Provider
      value={{
        timers,
      }}
    >
      {children}
    </TimerContext.Provider>
  );
}

export function useTimerProvider({
  task,
  onLocate,
}: {
  task: TaskDefinitionId;
  onLocate?: () => void;
}) {
  const timers = useContext(TimerContext)?.timers;
  onLocate && timers?.[task]?.setOnLocate(onLocate);
  return timers?.[task]?.component || <></>;
}
