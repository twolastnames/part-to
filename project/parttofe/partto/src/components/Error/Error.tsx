import React from "react";

import classes from "./Error.module.scss";
import { ErrorProps } from "./ErrorTypes";

export function Error({ code }: ErrorProps) {
  return (
    <div className={classes.error} data-testid="Error">
      <div className={classes.code}>{code}</div>
      <div className={classes.description}>
        This is bad. There is nothing here. You and your family could starve if
        you expected to cook dinner here. They could be saved by navigating to
        some place more useful in the site navigation menu.
      </div>
    </div>
  );
}
