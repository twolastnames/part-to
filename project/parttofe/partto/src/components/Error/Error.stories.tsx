import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Error } from "./Error";
import { ShellProvider } from "../../providers/ShellProvider";

const meta: Meta<typeof Error> = {
  component: Error,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Error>;

export const Simple: Story = {
  args: { code: 400 },
};
