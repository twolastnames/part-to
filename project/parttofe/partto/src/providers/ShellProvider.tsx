import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
