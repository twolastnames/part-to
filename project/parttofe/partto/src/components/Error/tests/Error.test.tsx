import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../providers/ShellProvider";
import { Error } from "../Error";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Error code={500} />
    </ShellProvider>,
  );
  const component = screen.getByTestId("Error");
  expect(component).toMatchSnapshot();
});
