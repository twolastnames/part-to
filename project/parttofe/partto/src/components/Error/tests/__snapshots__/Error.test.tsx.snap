// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="error"
  data-testid="Error"
>
  <div
    class="code"
  >
    500
  </div>
  <div
    class="description"
  >
    This is bad. There is nothing here. You and your family could starve if you expected to cook dinner here. They could be saved by navigating to some place more useful in the site navigation menu.
  </div>
</div>
`;
